sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs

```diff
@@ -0,0 +1,437 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+using Azure.Core.Pipeline;
+using Azure.Messaging.EventHubs.Consumer;
+using Azure.Messaging.EventHubs.Core;
+using Azure.Messaging.EventHubs.Processor;
+
+namespace Azure.Messaging.EventHubs.Primitives
+{
+    /// <summary>
+    ///   Provides a base for creating a custom processor for consuming events across all partitions of a given Event Hub
+    ///   within the scope of a specific consumer group.  The processor is capable of collaborating with other instances for
+    ///   the same Event Hub and consumer group pairing to share work by using a common storage platform to communicate.  Fault
+    ///   tolerance is also built-in, allowing the processor to be resilient in the face of errors.
+    /// </summary>
+    ///
+    /// <typeparam name="TPartition">The context of the partition for which an operation is being performed.</typeparam>
+    ///
+    public abstract class EventProcessor<TPartition> where TPartition : EventProcessorPartition, new()
+    {
+        /// <summary>Indicates whether or not this event processor is currently running.  Used only for mocking purposes.</summary>
+        private bool? _isRunningOverride;
+
+        /// <summary>
+        ///   The fully qualified Event Hubs namespace that the processor is associated with.  This is likely
+        ///   to be similar to <c>{yournamespace}.servicebus.windows.net</c>.
+        /// </summary>
+        ///
+        public string FullyQualifiedNamespace { get; }
+
+        /// <summary>
+        ///   The name of the Event Hub that the processor is connected to, specific to the
+        ///   Event Hubs namespace that contains it.
+        /// </summary>
+        ///
+        public string EventHubName { get; }
+
+        /// <summary>
+        ///   The name of the consumer group this event processor is associated with.  Events will be
+        ///   read only in the context of this group.
+        /// </summary>
+        ///
+        public string ConsumerGroup { get; }
+
+        /// <summary>
+        ///   A unique name used to identify this event processor.
+        /// </summary>
+        ///
+        public string Identifier { get; }
+
+        /// <summary>
+        ///   Indicates whether or not this event processor is currently running.
+        /// </summary>
+        ///
+        public bool IsRunning
```

Comment:
We figured there's a much higher number of callers that will be reading that property than there are those mocking it and that the mocking would be a small scope for usage/testing.

Given the performance difference between virtual and non-virtual properties, we didn't feel there was justification to make it fully virtual.   Are there scenarios that you're thinking of where the "once you set it, you own the value" approach that we've taken doesn't work?

Reply:
> Given the performance difference between virtual and non-virtual properties, we didn't feel there was justification to make it fully virtual. Are there scenarios that you're thinking of where the "once you set it, you own the value" approach that we've taken doesn't work?

I couldn't think of any.  It just felt like if we were finding the need to mock this - perhaps our customers would as well and it was unclear to me how they'd be able to play the same games in a way we were comfortable with.  I'm happy to keep it as is - especially since we could consider making it virtual in the future if we needed.

Reply:
Fair enough.   We added the setter for our own test mocking and use that exclusively.   So, for what its worth, we haven't found any gaps within our own test bed where that doesn't allow for test mocking.